logo.js

```diff
@@ -2379,32 +2523,67 @@ function LogoInterpreter(turtle, stream, savehook)
   });
 
   def("repeat", function(count, statements) {
-    count = aexpr(count);
-    statements = reparse(lexpr(statements));
-    for (var i = 1; i <= count; ++i) {
-      var old_repcount = self.repcount;
-      self.repcount = i;
-      try {
-        self.execute(statements);
-      } finally {
-        self.repcount = old_repcount;
+    return new Promise(function (resolve, reject) {
+      count = aexpr(count);
+      statements = reparse(lexpr(statements));
+      var i = 1;
+      function runLoop() {
+        while (i <= count) {
+          var old_repcount = self.repcount;
+          self.repcount = i;
+          i++;
+          var result;
+          try {
+            result = self.execute(statements);
+          } finally {
+            self.repcount = old_repcount;
+          }
+          if (isPromise(result)) {
+            result.then(runLoop).catch(function (err) {
+              i = count;
+              reject(err);
+            });
+            break;
+          }
+        }
+        if (i <= count) {
+          resolve();
+        }
       }
-    }
+      runLoop();
+    });
   });
 
   def("forever", function(statements) {
     statements = reparse(lexpr(statements));
-    for (var i = 1; true; ++i) {
-      var old_repcount = self.repcount;
-      self.repcount = i;
-      try {
-        self.execute(statements);
-      } finally {
-        self.repcount = old_repcount;
+    return new Promise(function (resolve, reject) {
+      var stop = false;
+      var i = 1;
+      function runLoop() {
```

Comment:
ISTM guts of this should be as simple as `return self.execute(statement).then(runLoop);`